File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Maldives.m
//
//  UIImage+Maldives.m
//  CountriesFlags
//

#import "Countries/UIImage+Maldives.h"

@implementation UIImage (Maldives)

+ (UIImage *)flagMaldivesWithSize:(CGSize)size {
    UIColor* colorRed = [UIColor colorWithRed:200.0f/255.0f
                                        green:16.0f/255.0f
                                         blue:46.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:132.0f/255.0f
                                           blue:61.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorGreen setFill];
    CGRect greenRect = CGRectMake(size.width / 72.0f * 12.0f,
                                  size.height / 48.0f * 12.0f,
                                  size.width / 72.0f * 48.0f,
                                  size.height / 48.0f * 24.0f);
    UIRectFill(greenRect);
    
    [colorWhite setFill];
    CGFloat radius = size.height / 48.0f * 8.0f;
    CGPoint outerCenter = CGPointMake(size.width / 72.0f * 39.0f, size.height / 2.0f);
    CGPoint innerCenter = CGPointMake(size.width / 72.0f * 42.0f, size.height / 2.0f);
    [[UIBezierPath bezierPathWithArcCenter:outerCenter
                                    radius:radius
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0
                                 clockwise:YES] fill];
    [colorGreen setFill];
    [[UIBezierPath bezierPathWithArcCenter:innerCenter
                                    radius:radius
                                startAngle:0.0f
                                  endAngle:M_PI * 2.0
                                 clockwise:YES] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Maldives.swift
//
//  UIImage+Maldives.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_the_Maldives "wikipedia"
     
     Generates the flag of the Republic of Maldives.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the Republic of Maldives.
     */
    internal static func flagMaldives(with size: CGSize) -> UIImage {
        let colorRed = UIColor(red: 200.0 / 255.0,
                               green: 16.0 / 255.0,
                               blue: 46.0 / 255.0,
                               alpha: 1.0)
        let colorGreen = UIColor(red: 0.0,
                                 green: 132.0 / 255.0,
                                 blue: 61.0 / 255.0,
                                 alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, true, 0.0)

        colorRed.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorGreen.setFill()
        let greenRect = CGRect(origin: CGPoint(x: size.width / 72.0 * 12.0, y: size.height / 48.0 * 12.0),
                               size: CGSize(width: size.width / 72.0 * 48.0, height: size.height / 48.0 * 24.0))
        UIRectFill(greenRect)

        colorWhite.setFill()
        let radius = size.height / 48.0 * 8.0
        let outerCenter = CGPoint(x: size.width / 72.0 * 39.0, y: size.height / 2.0)
        let innerCenter = CGPoint(x: size.width / 72.0 * 42.0, y: size.height / 2.0)
        UIBezierPath(arcCenter: outerCenter,
                     radius: radius,
                     startAngle: 0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()
        colorGreen.setFill()
        UIBezierPath(arcCenter: innerCenter,
                     radius: radius,
                     startAngle: 0,
                     endAngle: .pi * 2.0,
                     clockwise: true).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image ?? UIImage()
    }
}

#endif
